openedx/core/djangoapps/oauth_dispatch/docs/decisions/0006-enforce-scopes-in-LMS-APIs.rst

```diff
@@ -0,0 +1,225 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+(**Note to MS:**  Some details here are different from the implementation
+in your PR. See further explanations in the Consequences section below.)
+
+* In order to allow Applications to access data for their own
+  organization without inadvertently or maliciously gaining access
+  to data for other organizations, Applications need to be
+  linked to their own organization. To support this, open edX 
+  operators can configure Application-specific "available organizations",
+  which are akin to Application-specific "available scopes".
+
+* Introduce a new data model that associates available organizations
+  with DOT Applications.
+
+* The new data model will have a Foreign Key to the Organization_ table.
+  It will essentially be a many-to-many relationship between Organizations
+  and DOT Applications.
+
+* The organization associated with the Application will be included
+  in the JWT tokens requested by the Application.
+
+  * JwtBuilder_'s *build_token* functionality will be extended to include
+    the organization value in the token's payload. This payload is
+    cryptographically signed and so binds and limits the scopes in the
+    token to the organization.
+
+  * Since the organization value is in the token, any relying parties
+    that receive the token (including microservices) will be able to
+    enforce the scopes as limited to the organization.
+
+* **Question:** Should we distinguish between user_org and course_org?  
+  The former is the Enterprise relationship while the latter is the
+  content provider relationship.  In the future, we will also have
+  sponsorship relationship. This depends on whether large organizations
+  will want to compartmentalize their Applications.
+
+.. _Organization: https://github.com/edx/edx-organizations/blob/fa137881be9b7d330062bc32655a00c68635cfed/organizations/models.py#L14
+
+5. Introduce a new Permission class to enforce scopes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* The new `custom Permission`_ class will extend DOT's TokenHasScope_
+  Permission class.
+
+* The TokenHasScope_ permission allows API endpoints to declare the
+  scopes that they require in a *required_scopes* class variable.
+
+* The permission class will verify that the scopes in the provided JWT
+  are a proper superset of the scopes required by the requested view.
+
+* For now, the permission class will skip this verification if the
+  application is not a Restricted Application or if the token_type
+  was not a JWT token.
+
+  * **Question:** This will be an issue when microservices want to verify
+    scopes. Determining whether an access token is associated with a 
+    Restricted Application is an LMS-specific capability. Given this, how 
+    can we roll this out eventually to microservices? Do we need to
+    include a field in the token that indicates whether it was issued
+    to a Restricted Application?
```

Comment:
Could we not use the strategy of:  If scopes are provided in the JWT they get enforced whether or not the application is a Restricted Application.

Will applications that are not a Restricted Application have scopes in their JWTs?  Would it be invalid to enforce those scopes?

Reply:
We can't dothis because we do currently have scopes they are just not the ones we want.  Look at the examples in decision 0003.